streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -648,71 +794,142 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
+        while (!newTasks.isEmpty()) {
+            if (unfilledConsumers.isEmpty()) {
+                throw new IllegalStateException("Some tasks could not be distributed");
+            }
 
-                final Map<TaskId, Set<TopicPartition>> standbyTasks = new HashMap<>();
-                for (final TaskId taskId : clientMetadata.state.prevStandbyTasks()) {
-                    standbyTasks.put(taskId, partitionsForTask.get(taskId));
+            final Iterator<String> consumerIt = unfilledConsumers.iterator();
+            while (consumerIt.hasNext()) {
+                final String consumer = consumerIt.next();
+                final List<TaskId> consumerAssignment = assignments.get(consumer);
+                final TaskId task = newTasks.poll();
+                if (task == null) {
+                    break;
                 }
 
-                assignment.put(consumerId, new Assignment(
-                    assignedPartitions,
-                    new AssignmentInfo(
-                        minUserMetadataVersion,
-                        activeTasks,
-                        standbyTasks,
-                        partitionsByHostState,
-                        0)
-                        .encode()
-                ));
+                consumerAssignment.add(task);
+                if (consumerAssignment.size() == maxTasksPerClient) {
+                    consumerIt.remove();
+                }
             }
         }
 
-        // add empty assignment for "future version" clients (ie, empty version probing response)
-        for (final String consumerId : futureConsumers) {
-            assignment.put(consumerId, new Assignment(
-                Collections.emptyList(),
-                new AssignmentInfo().encode()
-            ));
-        }
+        return assignments;
+    }
 
-        return assignment;
+    /**
+     * Get the previous consumer for the partitions of a task
+     *
+     * @param taskPartitions the TopicPartitions for a single given task
+     * @param clientOwnedPartitions the partitions owned by all consumers in a client
+     * @param allOwnedPartitions all partitions claimed as owned by any consumer in any client
+     * @return the previous consumer that owned these partitions
+     *         empty string if it is a new task, or its owner is no longer in the group
+     *         null if its previous owner is from a different client or the task's partitions had different owners
+     */
+    String previousConsumerOfTaskPartitions(final Set<TopicPartition> taskPartitions,
+                                           final Map<TopicPartition, String> clientOwnedPartitions,
+                                           final Set<TopicPartition> allOwnedPartitions) {
+        String firstConsumer = "";
+        for (final TopicPartition tp : taskPartitions) {
+            // initialize the first consumer if this is the first iteration
+            if (firstConsumer != null && firstConsumer.equals("")) {
```

Comment:
Haha I knew this would get put down in the PR review but didn't get around to cleaning it up in time...thanks for the suggestion, that is much better 😄